CASM-3594: Add IMS checksum support to IUF manifest schema

jack-stanek-hpe opened the pull request:
## Summary and Scope
This change adds support to the IUF manifest schema to allow products to specify MD5 checksums for each IMS artifact. MD5 checksums are supported for IMS recipes, and for each of the rootfs, kernel, and initrd artifacts in IMS images. MD5 checksums will be verified by the `ims_upload` operation prior to upload.

I don't have much experience authoring JSON Schemas, so I apologize in advance if there are issues with this change!

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-3594](https://jira-pro.its.hpecorp.net:8443/browse/CASM-3594)

## Testing

### Tested on:

  * Local development environment

### Test description:

Load schema and run against test schemas including IMS artifacts with MD5 checksums.

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_


## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

haasken-hpe commented on src/api/models/iuf/iuf-manifest-schema.yaml:
```diff
@@ -399,19 +415,28 @@ properties:
-                  type: string
+                    The initrd file for the image.
+                  $ref: '#/$defs/IMSArtifact'
+          content_dirs:
```
A blank line before `content_dirs` would probably be helpful here for readability and consistency with the one added on line 391.

```suggestion

          content_dirs:
```

haasken-hpe commented on src/api/models/iuf/iuf-manifest-schema.yaml:
```diff
@@ -437,3 +462,21 @@ properties:
+$defs:
+  MD5Sum:
+    type: string
+    pattern: "^[0-9a-fA-F]+$"
```
are they always hex-encoded strings here?

jack-stanek-hpe replied:
Yes, `python-ims-helper` is using [hashlib.hash.hexdigest()](https://docs.python.org/3/library/hashlib.html?highlight=hexdigest#hashlib.hash.hexdigest) which returns a hexadecimal string. On second glance I think I need to remove the uppercase `[A-F]` character set, since MD5 validation is just a string comparison.

haasken-hpe commented on src/api/models/iuf/iuf-manifest-schema.yaml:
```diff
@@ -437,3 +462,21 @@ properties:
+          The relative path to the IMS artifact within its image's directory
+      md5sum:
+        description: >
+          The MD5 checksum of the IMS artifact
```
It's probably worth explicitly defining behavior if this is omitted, which I think is to do no verification.

jack-stanek-hpe replied:
I'll add a comment to this effect.

haasken-hpe reviewed (approved):
Let's bump the minor component of the version on line 37.

This change is technically not backwards compatible, but semver says that no assumptions about stability should be made about components with a major version of 0, so we're okay to do this. We will definitely want to communicate clearly about this in case any teams have started to use the old schema though.

jack-stanek-hpe commented:
@haasken-hpe, as I was addressing review comments I realized `template_dictionary` had the wrong type (`string` instead of `object` with string values), I have updated that as well. If you have a moment, I'd just like a quick check to make sure I've specified the types right.

dzou-hpe commented:
@jack-stanek-hpe you need to bump the `.version` file as well. once it is merged we need to tag master branch with that version so the new image is released

dzou-hpe reviewed (changes requested):
need .version bump

haasken-hpe commented on src/api/models/iuf/iuf-manifest-schema.yaml:
```diff
@@ -373,6 +373,32 @@ properties:
+                    Key-value pairs to be templated into the recipe when
+                    building the image.
+                  type: object
+                  properties: {}
```
I do not believe this is necessary. Testing:

```
C02ZH00ALVCJ: cray-nls (CASM-3594-update-schema *%) $ git diff
diff --git a/src/api/models/iuf/iuf-manifest-schema.yaml b/src/api/models/iuf/iuf-manifest-schema.yaml
index 264fd61..1b01b03 100644
--- a/src/api/models/iuf/iuf-manifest-schema.yaml
+++ b/src/api/models/iuf/iuf-manifest-schema.yaml
@@ -393,7 +393,6 @@ properties:
                     Key-value pairs to be templated into the recipe when
                     building the image.
                   type: object
-                  properties: {}
                   additionalProperties:
                     type: string
                 md5sum:
C02ZH00ALVCJ: cray-nls (CASM-3594-update-schema *%) $ go build
C02ZH00ALVCJ: cray-nls (CASM-3594-update-schema *%) $ cat examples/iuf-product-manifest/ims-artifacts.yaml
---
name: myproduct
version: 1.2.3
iuf_version: ^0.3.0

content:
  ims:
    recipes:
    - path: myrecipe
      recipe_type: kiwi-ng
      linux_distribution: sles15
      template_dictionary:
        fruit: apple
        vegetable: kale
      md5sum: d3b07384d113edec49eaa6238ad5ff00
C02ZH00ALVCJ: cray-nls (CASM-3594-update-schema *%) $ ./cray-nls validate examples/iuf-product-manifest/ims-artifacts.yaml
```

Is it also worth checking in an example file like the above? It's kind of helpful for validating schema as you're writing it, and maybe we could incorporate these into some tests?

jack-stanek-hpe replied:
Ah, good to see that this still works. I interpreted the `additionalProperties` description in the [core JSON Schema spec](https://json-schema.org/draft/2020-12/json-schema-core.html#additionalProperties) to mean that the `properties` attribute was necessary, but since there are no required attributes, I wanted it to explicitly be empty.
> The behavior of this keyword depends on the presence and annotation results of "properties" and "patternProperties" within the same schema object. Validation with "additionalProperties" applies only to the child values of instance names that do not appear in the annotation results of either "properties" or "patternProperties".

haasken-hpe reviewed (approved):
This looks good to me. One minor comment, but I don't think it makes a functional difference.

Of course, needs the `.version` update that Di mentions, and should be rebased on master and re-tested once I merge #82.

haasken-hpe commented:
When you rebase, please bump the `version` property in `iuf-manifest-schema.yaml` to `0.3.0` as well as the `.version` and Helm chart version updates.
